src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-subcomponent-interactor/table-subcomponent-interactor.component.html

```diff
@@ -0,0 +1,12 @@
+<div [ngClass]="displayTopLineClassExpanded(complex, i, j)"></div>
```

Comment:
Because of how angular works, methods call in `ngClass` and `ngIf` are called periodically even if nothing changes, which is a bit inefficient. One way to work around this is to call these methods in `ngOnChanges`, set the values to variables and then use the variables in the HTML.

For example, for this component, the ts file would be:
```
...
export class TableSubcomponentInteractorComponent implements OnChanges {
  @Input() complex: EnrichedComplex;
  @Input() i: number;
  @Input() j: number;
  @Input() enrichedInteractors: EnrichedInteractor[];

  displayTopLineClass: string;
  displayBottomLineClass: string;
  interactorInComplex = false;
  interactorStoichiometryText: string;
  interactorStoichiometryValue: string;
  interactorInSubComplex = false;
  subComplexInteractorStoichiometryText: string;
  subComplexInteractorStoichiometryValue: string;

  // findInteractorInComplex = findInteractorInComplex;
  // getStoichiometry = getStoichiometry;
  // stoichiometryOfInteractors = stoichiometryOfInteractors;

  ngOnChanges(changes: SimpleChanges): void {
    this.displayTopLineClass = this.displayTopLineClassExpanded(this.complex, this.i, this.j);
    this.interactorInComplex = !!findInteractorInComplex(this.complex.complex, this.el.identifier);
    this.interactorStoichiometryText = getStoichiometry(this.complex.complex, this.el.identifier);
    this.interactorStoichiometryValue = stoichiometryOfInteractors(this.complex.complex, this.el.identifier);
    this.interactorInSubComplex = !!this.findInteractorInExpandedSubComplex(this.interactor, this.complex.complex, this.el.identifier);
    this.subComplexInteractorStoichiometryText = this.getStoichiometryInExpandedSubComplex(this.interactor, this.el.identifier);
    this.subComplexInteractorStoichiometryValue = this.stoichiometryOfInteractorsExpandable(this.interactor, this.el.identifier);
    this.displayBottomLineClass = this.displayBottomLineClassExpanded(this.complex, this.i, this.j);
  }
...
```

Then, the HTML would be
```
<div [ngClass]="displayTopLineClass"></div>
<cp-table-interactor-stoichiometry
  *ngIf="interactorInComplex"
  [stoichiometryText]="interactorStoichiometryText"
  [stoichiometryValue]="interactorStoichiometryValue"
></cp-table-interactor-stoichiometry>
<cp-table-interactor-stoichiometry
  *ngIf="interactorInSubComplex"
  [stoichiometryText]="subComplexInteractorStoichiometryText"
  [stoichiometryValue]="subComplexInteractorStoichiometryValue"
></cp-table-interactor-stoichiometry>
<div [ngClass]="displayBottomLineClass"></div>
```

These changes mean that the methods are only called when there are changes on the inputs, and not periodically. These changes are not needed, but if you want to improve performance a bit, this type of change on different components could do that.